examples/sessions/custom_data_type/src/main.rs

```diff
@@ -0,0 +1,148 @@
+//! An introduction to storing and retrieving session data with a custom data type, in a type safe
+//! way, with the Gotham web framework.
+
+extern crate gotham;
+#[macro_use]
+extern crate gotham_derive;
+extern crate hyper;
+extern crate mime;
+extern crate serde;
+#[macro_use]
+extern crate serde_derive;
+extern crate time;
+
+use hyper::{Response, StatusCode};
+
+use gotham::http::response::create_response;
+use gotham::pipeline::new_pipeline;
+use gotham::pipeline::single::single_pipeline;
+use gotham::router::Router;
+use gotham::router::builder::*;
+use gotham::state::{FromState, State};
+use gotham::middleware::session::{NewSessionMiddleware, SessionData};
+
+// A custom type for storing data associated with the user's session.
+#[derive(Clone, Deserialize, Serialize, StateData)]
+struct VisitData {
+    count: usize,
+    last_visit: String,
+}
+
+/// Handler function for `GET` requests directed to `/`
+///
+/// Each request made will update state about your recent visits, and report it back.
+fn get_handler(mut state: State) -> (State, Response) {
+    let maybe_visit_data = {
+        let visit_data: &Option<VisitData> = SessionData::<Option<VisitData>>::borrow_from(&state);
+        visit_data.clone()
```

Comment:
I was trying to work out whether this could be done without the clone() and without the type-coercing noise. I think that deref() and deref_mut() help a lot here. This is what I came up with: 

```
fn get_handler(mut state: State) -> (State, Response) {
    let (visit_count, body) = match SessionData::<Option<VisitData>>::borrow_from(&state).deref() {
        &Some(ref visit_data) => (
            visit_data.count,
            format!(
                "You have visited this page {} time(s) before. Your last visit was {}.\n",
                visit_data.count, visit_data.last_visit,
            ),
        ),
        &None => (0, "You have never visited this page before.\n".to_owned()),
    };
    ...
    {
        *SessionData::<Option<VisitData>>::borrow_mut_from(&mut state).deref_mut() =
            Some(VisitData {
                count: visit_count + 1,
                last_visit: format!("{}", time::now().rfc3339()),
            });
    }
    (state, res)
}
```

I'm not sure if I like my match that returns a tuple of (visit_count, body) or not: I think my mind is still in python mode. I suspect that it would get increasingly awkward as the view gets more complicated. Might be better to stick with .clone().